R/utils.R

```diff
@@ -0,0 +1,146 @@
+# Implements "whitespace eating" functionality on top of a normal TextWriter.
+# (WSTextWriter doesn't inherit from TextWriter, as it presents a different
+# interface.)
+#
+# WSTextWriter relies on the caller distinguishing between writes of important
+# content, and writes of whitespace that may or may not be elided (`.$write()`
+# vs `.$writeWS()`).
+#
+# At any point, `eatWS` may be called, which will cause any recent `writeWS`
+# operations (i.e. those since either the beginning of time, or the most recent
+# `write` operation) to be undone, AND for any future `writeWS` calls to be
+# ignored. A call to `write` will be respected, and will restore normal
+# behavior.
+WSTextWriter <- R6Class("WSTextWriter",
+  private = list(
+    writer = "TextWriter"
+  ),
+  public = list(
+    initialize = function() {
+      private$writer <- TextWriter$new()
+    },
+    close = function() {
+      private$writer$close()
+    },
+    write = function(text) {
+      # If an error is going to happen while evaluating `text`, let's make it
+      # happen before we mutate any of our internal state
+      force(text)
+
+      # Stop eating whitespace
```

Comment:
Note to self: Nit: I think this comment is misleading. `setSuppress` blocks _all_ writes. We're doing this to restore writing functionality globally on the writer because this interface opts never to suppress a non-whitespace write.

Reply:
Do we really need this suppress behavior? Is it used by other callers? If not, it seems a little more intuitive IMHO to have this whitespace-suppressing writer manage whether or not it wants to write WS and then just omit the call to `write` in `writeWS` if it's in suppressing mode. That simplifies the interface on the textWriter, right?